### PR TITLE
Enhancement: add collective influence measure

### PR DESCRIPTION
**Pull request by smopucilowski**
A new paper by Morone & Makse appearing in Nature a few days ago (doi:10.1038/nature14604) introduced the concept of Collective Influence. Briefy: this node measure quantifies the influence of a node by finding an optimal (minimal) set of structural influencers. Nodes with high collective influence, when removed, break down the network into many disconnected pieces.

This pull request implements their node measure.

### CODE REVIEW

File: networkx/algorithms/centrality/collective_influence.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+      [2] István A. Kovács and Albert-László Barabási:
+      Network science: Destruction perfected
+      Nature 524, 38–39 (06 August 2015)
+      doi: 10.1038/524038a
+    """
+
+    reduced_degree = lambda node: G.degree(node)-1
+
+    if u is None:
+        nodes = G.nodes_iter()
```
Comment (MridulS) [inline networkx/algorithms/centrality/collective_influence.py]: Can you change this to G.nodes(). nodes() method will return an iterator after https://github.com/networkx/networkx/pull/1716 is merged.

Comment (smopucilowski) [inline networkx/algorithms/centrality/collective_influence.py]: Thank you. I have committed the change.

**Comment by jfinkels** [comment]
How about this, which only makes a single call to the all-pairs shortest path algorithm, instead of multiple calls to the single-source shortest path algorithm:

    def collective_influence(G, u=None, distance=2, _paths=None):
        if u is None:
            if _paths is None:
                _paths = nx.all_pairs_shortest_path(G)
            # TODO This can be trivially parallelized.
            return {v: collective_influence(G, v, distance, _paths[v]) for v in G}
        if _paths is None:
            _paths = nx.single_source_shortest_path(G, source=u, cutoff=distance)
        return _collective_influence(G, u, distance, _paths)


    # Pre-condition: `paths` is not empty.
    def _collective_influence(G, u, distance, paths):
        reduced_degree = lambda node: G.degree(node) - 1
        frontier = (v for v, path in paths.items() if len(path) - 1 == distance)
        return reduced_degree(u) * sum(map(reduced_degree, frontier))

**Comment by dschult** [comment]
I like the original submission better @jfinkels because the indentation structure shows its process. Also, helper functions that only get used once are usually better in-line. Finally the original code is closer to the paper's explanation. But I'm fine with changing it too if @smopucilowski likes the suggested approach.

**Comment by jfinkels** [comment]
> Also, helper functions that only get used once are usually better in-line.

Oops, you're right of course, I originally had a slightly different implementation in mind, then changed my mind halfway through writing it.

> Finally the original code is closer to the paper's explanation.

Makes sense to me, I was mostly just concerned about the cost of repeatedly calling single-source shortest paths when a single call to all-pairs shortest paths would suffice.

**Comment by hagberg** [comment]
If the cutoff is small the single_source implementation might be faster too.

**Comment by smopucilowski** [comment]
I'm happy to include a change which precomputes shortest paths, however, for large networks the creation of this dictionary risks consuming all memory.

Please advise which approach suits networkx best.

**Comment by smopucilowski** [comment]
Excellent suggestion @dschult. I have updated my branch with it.

**Comment by dschult** [comment]
Only other idea/suggestion for efficiency I have is to consider whether storing the reduced degrees in a dict might be more efficient than repeatedly calling G.degree(node). Something like:

    reduced_degree = {n:d-1 for n,d in G.degree()}  # Note: github master has G.degree() an iterator

Then later ```reduced_degree[node]``` replaces reduced_degree(node). Also you would have to remove the ```map()``` by putting ```reduced_degree[node]``` inside the frontier_nodes set comprehension. If you move the ```sum()``` to the line with the set comprehension you can save having to create the set.

    frontier_sum = sum(reduced_degree[node] for node, path_length in path_lengths.items()
                                    if path_length == distance )

Not a big deal if this is too much--I'm just in the mode of working with iterators lately. It's your PR.

Finally, you should add this change to the doc/source/reference/release_2.0.rst and consider adding
yourself to the credits.rst file in that directory.

**Comment by smopucilowski** [comment]
Thank you for your suggestions @dschult. I have added this change to the documentation as appropriate.

Regarding storing reduced degrees in a dictionary, I had considered using some sort of dictionary to store these values. However,
* G.degree(node) is a dictionary lookup already within the graph class, which has the same computational requirements as a dictionary lookup in our own structure.
* If the function is used to query a single node, readability of the code suffers as it gets messy: we only wish to calculate the reduced degree of local nodes, not the entire graph.

What are your thoughts?

**Comment by dschult** [comment]
I think you have to compute the reduced degree of all nodes when you find the influence of all nodes. But when computing for only one node you are right--we only need it for local nodes.

The G.degree(node) vs reduced_degree[node] question is best solved by testing... evaluating G.degree itself takes time, and then inside the function there's a lookup and a function call to len. Could you try changing the reduced_degree function to look it up in a dict? I use IPython for timing with the %timeit "magic" command.

    rdeg = {n:d-1 for n, d in G.degree()}
    reduced_degree = lambda node: rdeg[node]

This should be slower for individual nodes, but might be faster for full graphs.  Again, if this is too much let me know and we'll mess with that issue as part of the iterator switch over of the shortest_path routines.

**Pull request closed without merging**